artifact add: default to latest tag when no tag specified

shiavm006 opened the pull request:
Fixes issue where podman artifact add did not default to :latest tag like container images do, causing inconsistent behavior.

fixes #27083

```release-note
None
```

Honny1 reviewed (changes requested):
Thanks, I've checked the code. I believe the "latest" tag should be baked into the artifact by default.

---
You can retrigger Cirrus CI jobs. Just open the details of a failed task (for example, at its URL ). If you are authenticated to Cirrus CI with your GitHub account, you should see a "Rerun" button. For Packit jobs, you can rerun them from a comment. [See the documentation on retriggering.](https://packit.dev/docs/retriggering). Please remove the commits that were made solely to retrigger the CI pipeline.

Honny1 reviewed (commented):
LGTM, once CI passes.

shiavm006 commented on pkg/libartifact/store/store.go:
```diff
@@ -217,6 +224,11 @@ func (as ArtifactStore) Add(ctx context.Context, dest string, artifactBlobs []en
 
+	// Add :latest tag if no tag/digest is present
+	if !strings.Contains(dest, ":") && !strings.Contains(dest, "@") {
+		dest = dest + ":latest"
```
implemented teh changes

shiavm006 commented:
> Out of curiosity, how did you create this commit?

I made the changes locally, ran golangci-lint to catch the style issue, then staged the fix and used git commit --amend -s to squash it into the existing commit with DCO, followed by git push --force-with-lease to update the same branch on my fork.

shiavm006 commented:
> @shiavm006 It looks like your commit message has incorrect formatting. You can use the --amend flag to fix the text. I think fixing the escaped newlines should help. Out of curiosity, how did you create this commit?

I’ve amended the commit message to proper subject/body (no escaped newlines) and kept the DCO. Also applied the += ":latest" nit and squashed into the same commit. Waiting on CI now.

Honny1 commented:
@shiavm006 Could you please retrigger the 'Build for MacOS' job? It looks like it failed due to a timeout.

Honny1 commented:
@shiavm006 Please update the tests to expect the `:latest` tag.

Honny1 commented on pkg/libartifact/store/store.go:
```diff
@@ -132,7 +132,14 @@ func (as ArtifactStore) Inspect(ctx context.Context, nameOrDigest string) (*liba
+
+	// Normalize name to :latest if no tag/digest present for single lookup
+	lookupName := nameOrDigest
+	if !strings.Contains(nameOrDigest, ":") && !strings.Contains(nameOrDigest, "@") {
```
The test for `podman artifact inspect with full or partial digest` fails because `:latest` is being added. I think we should handle the `:latest` tag inside `artifacts.GetByNameOrDigest` instead.

Honny1 commented on test/e2e/artifact_test.go:
```diff
@@ -204,7 +223,7 @@ var _ = Describe("Podman artifact", func() {
 		rmFail := podmanTest.Podman([]string{"artifact", "rm", "foobar"})
 		rmFail.WaitWithDefaultTimeout()
-		Expect(rmFail).Should(ExitWithError(125, fmt.Sprintf("Error: %s: artifact does not exist", "foobar")))
+		Expect(rmFail).Should(ExitWithError(125, fmt.Sprintf("Error: %s: artifact does not exist", "foobar:latest")))
```
The error message is missing the `:latest` tag.

Honny1 reviewed (changes requested):
Based on the test results, the `:latest` tag should be resolved during the lookup, not before it. We also need to consider backward compatibility.

shiavm006 committed:
artifact: default untagged artifacts to :latest tag (Fixes #27083)

shiavm006 committed:
artifact: fix :latest tag handling for backward compatibility

Store artifacts with :latest tag when no tag specified.

Update GetByNameOrDigest to handle both original name and :latest lookups.

Maintain backward compatibility for existing artifacts.

Fix error messages to include :latest when appropriate.

Fixes #27083

Signed-off-by: shiavm006 <shivammittal42006@gmail.com>

rhatdan commented:
Why was this closed, this is still needed?
